Add plots, statistics and basic recommendations

Alexander-Serov opened the pull request.

Alexander-Serov committed:
Allow to get a simple version of all ETFs' history and plot

Alexander-Serov committed:
Provide History class that locally stores daily price history and can update online

Alexander-Serov committed:
Update local table format

Alexander-Serov committed:
Allow to make a simple recommendation

Alexander-Serov committed:
Allow simple recommendations 2

Alexander-Serov committed:
Add colors to simple strategy

Alexander-Serov committed:
Hotfix price update when the market is open. Should fix #4.

Alexander-Serov committed:
Allow to log warning to file instead of showing error messages.

Alexander-Serov committed:
Add a recommendation based on daily change and not weekly change.

Alexander-Serov committed:
Fix obsolete figis. Output a table with main stats.

Alexander-Serov committed:
For years, determine the range by a high quantile rather than maximum. Same for minimum. Keep real extrema for weeks. Should increase the statistical robustness of the estimates, while making sure there are no incorrect decisions on the week scale.

Alexander-Serov committed:
Move and better name logs.

sserov commented on .gitignore:
```diff
@@ -1,3 +1,5 @@
+token
```
Okay, all utility files and folders as well as data files should be ignored

Alexander-Serov replied:
So it's OK for you or something is missing?

Alexander-Serov replied:
Так что-то еще хочешь добавить или как?

sserov replied:
Нет, всё окей

sserov replied:
По-моему, тут была суть в том, что нужно еще временные файлы туда засунуть, а не только токен

sserov replied:
Например, у меня ПайШарм создает всякую пуфню

Alexander-Serov replied:
Ну можешь предложить изменение гитигнора.

Alexander-Serov replied:
Ну так?

sserov replied:
```suggestion
token
.idea/
```

sserov commented on README.md:
```diff
@@ -1,2 +1,17 @@
 # tinkoff-analysis
-A solution using tinkoff-api investments for market situation analysis
+A solution using `tinkoff-api` investments for market situation analysis
```
```suggestion
A solution using `tinkoff-api` investments and [`open-api-python-client`](https://github.com/Awethon/open-api-python-client) python client for market situation analysis
```

sserov commented on main-analysis.py:
```diff
@@ -0,0 +1,3 @@
+"""
+An executable for running future analysis from the terminal
+"""
```
No newline at the end

Alexander-Serov replied:
Поправлю

sserov commented on main-recommend.py:
```diff
@@ -0,0 +1,9 @@
+File allowing to access investment recommendation from the terminal
+"""
+
+from analysis import History
```
Simple interface to the recommendations, ok

Alexander-Serov replied:
К чему этот комментарий?

sserov replied:
Думаю, это был просто комментарий

Alexander-Serov replied:
Странный абстрактный комментарий.￼

sserov commented on requirements.txt:
```diff
@@ -0,0 +1,3 @@
+pandas
+matplotlib
+colorama
```
```suggestion
colorama
numpy
pytz
```

Need to include also

```
openapi_client
openapi_genclient
```
?

Alexander-Serov replied:
If we add those two, pip install will fail. I don't think they are on pip.
I think they should rather be in the readme.

sserov replied:
Вроде, пишут, что через pip можно с pypi. Только ссылку надо будет как-то указать.

https://github.com/Awethon/open-api-python-client

sserov replied:
@Alexander-Serov ?

Alexander-Serov replied:
Не, это все фигня. Не сможем мы объяснить людям, которые захотят использовать, что нужно установить вот как-то сложно. Поэтому хорошо, если мы перейдем #22 в какой-то момент.

sserov replied:
Окей, тогда оставим как ишшу

sserov commented on main-analysis.ipynb:
```diff
@@ -0,0 +1,435 @@
+{
```
Simple examples and plots. There are some strange leaps from 0 to 80% in FXWO (?) and some other tickers. Visualization with all the tickers in one plot is rather hard to read. Maybe we should use Plotly in future.

Alexander-Serov replied:
Комментарий понял, да, там есть неожиданные скачки, но там такие данные грязноватые. Уверен, что Талиб сразу подавится. 
Но что конкретно ты предлагаешь тут?

sserov replied:
Странно, что данные грязные. Надо разобраться, в чем проблема у них

Alexander-Serov replied:
Так что ты предлагаешь?

Alexander-Serov replied:
Есть новости на эту тему после твоих проб?

sserov replied:
Проб? Нет, это не смотрел, фиксил другое. Вообще, странно, что данные грязные (точно грязные)? Потому что то, что Пуфи смотрел, выглядело более-менее. Если реально грязные, то нужно будет, видимо, сглаживание делать. Об этом ишшу #7

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+import inspect
+import time
+
+from openapi_client import openapi
```
A lot of unused imports

Alexander-Serov replied:
Починил.

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+
+colorama.init()
+
+# Load the token
```
There is a section outside the functions. Maybe it's better to move it into init function or constructor of general class

Alexander-Serov replied:
предполагалось, что эта секция будет использоваться разными классами. это просто врапперы вокруг функций из апи. можно перенести в отдельный файл если тебе тут не нравятся?

sserov replied:
Смущало, вроде, что при простом импорте уже выполняется много действий и в том числе создаются папки

Alexander-Serov replied:
Я думаю папки создавались для того чтобы хранить кэш. Но они могут создаваться при запросе истории с сервера, это правда.

Alexander-Serov replied:
Починено в #21 ? Закрываем тут?

sserov replied:
Да, это всё починено

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+
+def get_all_etfs(reload=False):
+    if get_all_etfs.etfs is None or reload:
+        get_all_etfs.etfs = market.market_etfs_get().payload.instruments
```
Global variables in the module is not a good practice. Suggest to wrap it into a class

sserov replied:
Function members are also not needed here. Suggest to make the function a class method which will modify class members

Alexander-Serov replied:
хорошо, согласен. переделаем это в класс.

Alexander-Serov replied:
Но опять же:
- предполагалось, что эти функции будут использоваться и из других классов.
- если будем менять, может в отдельной ишшу?

sserov replied:
Ничего не мешает обращаться из других классов к объекту этого класса напрямую или через геттеры/сеттеры. Но всяко лучше, чем к переменной самого модуля, как по мне.

Возможно, в отдельной ишшу

Alexander-Serov replied:
Тоже перенес в #21 ?

sserov replied:
Закрыто в #21

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+        #                                  to=end.isoformat(), interval=interval))
+        candles = hist.payload.candles
+        candles_dict = [candles[i].to_dict() for i in range(len(candles))]
+        df = pd.DataFrame.from_dict(candles_dict)
```
from_dict gets list instead of dict

Alexander-Serov replied:
правда странно.

sserov replied:
Ага

sserov replied:
@Alexander-Serov ?

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+        candles_dict = [candles[i].to_dict() for i in range(len(candles))]
+        df = pd.DataFrame.from_dict(candles_dict)
+    except Exception as e:
+        if figi not in obsolete_tickers.values():
```
What's the problem with obsolete tickers?

Alexander-Serov replied:
это те, которые перестали торговаться: существуют в истории, но когда пытаешься запросить недавнюю историю, возникает ошибка.

sserov replied:
А, окей

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+    return df
+
+
+@lru_cache(maxsize=1000)
```
Cache?

Alexander-Serov replied:
Да. В чем вопрос?

sserov replied:
Вопрос, видимо, был в том, что за кэш. Ну, допустим, он нужен

Alexander-Serov replied:
Кэш автоматически сохраняет результат функции по запросу в виде словаря. Для функции как здесь соответствие фиги и тикера - это в самый раз так как оно не меняется.

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+def get_ticker_for_figi(figi):
+    if figi in obsolete_tickers.values():
+        return None
+    try:
```
Why here there is exceptions processing but in the previous function there isn't one? What's the puffnya with obsolete tickers?

Alexander-Serov replied:
потому что предыдущая функция делает обратную операцию. в текущем коде я только запрашивал тиккер по фиджи и не наоборот.

sserov replied:
Окей

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+                         weeks=52),
+                     interval='month'):
+    """
+    Get history _data with a given interval for all available ETFs.
```
Missing parameters in docstring

Alexander-Serov replied:
починено

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+                             f'ticker={ticker} and period=[{start}, {end}]')
+
+        one_etf_history.drop(columns=['interval'], inplace=True)
+        if 'ticker' not in one_etf_history.columns:
```
How there can be different sets of columns in returned dataframes?

Alexander-Serov replied:
честно, не могу тебе уже сейчас сказать.

sserov replied:
Странная строчка, ну ладно

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+
+def get_etfs_history(end=dt.datetime.now(dt.timezone.utc),
+                     start=dt.datetime.now(dt.timezone.utc) - dt.timedelta(
+                         weeks=52),
```
A strange end of line

Alexander-Serov replied:
поправим.

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+        if 'ticker' not in one_etf_history.columns:
+            one_etf_history['ticker'] = ticker
+
+        # Merge into the large table
```
Not merge but concat?

Alexander-Serov replied:
Да, это правда что должно работать. Предлагаю создать ишшу и в следуюущей сделать. Пока ведь работает уже.

sserov replied:
Вроде, тут вопрос был в том, что там и так append, но просто в комменте стоит merge

Alexander-Serov replied:
Имелся в виду мерж Потому что даты могут перекрываться. ￼

Alexander-Serov replied:
?

sserov replied:
Так он закомменчен и изменен в итоге на append?

Alexander-Serov replied:
Сейчас точно аппенд. И да, в последнем коммите я заменил на аппенд. Ты ведь именно этого просил?

Вот только, хоть убей, не помню, как я это делал.

sserov replied:
Я просил комментарий поменять на аппенд, раз уж он написан

Alexander-Serov replied:
Ага, ну хорошо.

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+        length = interval_dt
+
+    one_period = dt.timedelta(weeks=52)
+    need_divide = length > one_period
```
Unused variable need divide

Alexander-Serov replied:
починил

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+    Close price when the market is close was verified.
+    #todo verify current price when the market is open
+
+    Parameters
```
Different format of docstring?

Alexander-Serov replied:
Да, давай пользоваться этим. Это Нумпай формат и его легче читать.

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+            f'Ticker and figi point to different products: {figi} {get_figi_for_ticker(ticker)}')
+
+    if figi in obsolete_tickers.values() or figi is None:
+        return np.nan
```
Why np.nan for obsolete tickers? They cannot be processed at all, can they?

Alexander-Serov replied:
в смысле предлагаешь ошибку?

sserov replied:
Предлагаю возвращать тот же тип, но только пустой. Или в худшем случае None

Alexander-Serov replied:
Ну не знаю, цена должна быть float. Не известно цена это нан, как обычно￼. Ноне используется обычно для классов нечисленных мне кажется.

Alexander-Serov replied:
?

sserov replied:
Ну это имеет разные смыслы. None -- объекта нет, а np.nan -- нечисловое значение. Но да ладно, пусть так

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+    data is always updated.
+
+    Usage:
+    History.update() # to update the local database of prices
```
Meaning an instance of the class here?

Alexander-Serov replied:
да, починил.

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+
+            # Drop nans in time stamps
+            l1 = len(self._data)
+            self._data.drop(self._data[pd.isna(self._data.time)].index, inplace=True)
```
For what should we store rows with nan timestamps and drop them during each load? Maybe it's better to kill them before saving (the are already dropped in ```def get_etfs_daily_history```)

Alexander-Serov replied:
я думаю это файлсэйф: если они вдруг нечаянно записались то потом корректно их загружать. 
это правда что может теперь уже и не нужна эта строчка.

sserov replied:
Да, возможно, стоит выкинуть

Alexander-Serov replied:
Ну твое возможно мне особо ничего не говорит: ты либо предлагаешь, либо мы об этом не думаем.￼

Alexander-Serov replied:
?

sserov replied:
Вроде, это проверяется в def get_etfs_daily_history, и поэтому каждый раз выбрасывать их при загрузке смысла нет. Предлагается эту проверку удалить

Alexander-Serov replied:
Я могу удалит, но предлаегается тебе сначала написат тест, который будет проверят, что ничего не сломается, когда я удалит.

Alexander-Serov replied:
Можешь его прямо в ветку коммитнуть.

sserov replied:
Не, давай забьем на тест. Пока не до него

Alexander-Serov replied:
Вот видишь. Не все так просто выкидывать. Тем более работает...

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+            if len(self._data) < l1:
+                log_to_file(f'{l1 - len(self._data)} NaN time stamps dropped.')
+
+            self._data.time = pd.to_datetime(self._data.time)
```
Can be moved into read_csv (see ```parse_dates``` parameter in https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html)

Alexander-Serov replied:
надо будет сделать, убедившись что все работает.

Alexander-Serov replied:
ты в принципе можешь тоже что-то сделать.

sserov replied:
Да, надо будет сделать

sserov replied:
Я сделать могу, но я предпочитаю сначала знать, что делать, а не просто какие-то мелкие дырки закрывать подряд

Alexander-Serov replied:
Ну теперь ты знаешь. Можешь либо сделать либо ищу создать￼

Alexander-Serov replied:
Ну так что?

sserov replied:
#23

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+                try:
+                    os.rename(tmp_file, file)
+                except Exception as e:
+                    success = False
```
```suggestion
                    print(f"Unable to rename {tmp_file} into {file}")
                    success = False
```

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+                    success = False
+                    raise e
+
+        return success
```
```success = False``` are not needed because exceptions are raised in these cases

Alexander-Serov replied:
это правда. надо будет поправить.

Alexander-Serov replied:
что с этим делаем в итоге?

sserov replied:
Убирай этот success. Даже пайшарм говорит, что она бессмысленная

Alexander-Serov replied:
мой не говорит.

Alexander-Serov replied:
но хорошо.

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+                f'The received data frame cannot be empty because one of the '
+                f'dates was already present in the history data. Requested dates: {start_date} -- {today}.')
+
+        self._tickers = tickers
```
And why do we update the tickers list in all cases even if reload is not needed?

Alexander-Serov replied:
ЕТФ торгующиеся на рынке могли поменяться. Релоад = фальш говорит лишь о том что он будет использовать кэшированные данные если они доступны. Иначе все запросит заново.

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+        else:
+            """Drop all data from the original table that repeats in the new 
+            one """
+            old_data = copy.deepcopy(self._data)
```
Why not to make ```pd.concat``` and then ```drop_duplicates```? It also should not spoil date columns in theory

Alexander-Serov replied:
на самом деле нет, с тех пор, я нашел еще лучше функцию. 
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.combine_first.html#pandas.DataFrame.combine_first
Может как раз это то, что нужно?

sserov replied:
Кажись, да, это то, что нужно

Alexander-Serov replied:
Ну опять же, может и то, что нужно. Но текущая версия уже работает. И писалась год назад. Я считаю что пока можно не обращать на это внимания: если я просто так ее сюда вставлю без теста может сломаться что-то еще. Тут надо сначала тест писать.

sserov replied:
Окей, ну это и был коммент год назад про то, что (возможно) стоит улучшить. Если так работает, то, думаю, можно и оставить

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+                                                filter_52w, ['figi', 'h']].groupby(
+                          by='figi').quantile(max_quantile)['h'].to_dict(),
+                      'max_52w-10%': (self._data.loc[self._data.time >=
+                                                     filter_52w, ['figi', 'h']].groupby(
```
Strange identation

Alexander-Serov replied:
починил.

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+        are still kept. In theory, a similar procedure can be implemented, but higher resolution
+        data are required.
+
+        :param position what time of day (o, c, h, l) to use to
```
What are o, c, h, l?

Alexander-Serov replied:
добавил информацию.

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+
+    def update(self, reload=False, verbose=False):
+        """
+        Fetch the latest _data from server starting from 2 last known days.
```
From 2 last known days?

Alexander-Serov replied:
устарело. исправил.

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+                          self._data.loc[self._data.time >= filter_1w, ['figi', position]].groupby(
+                              by='figi').first()[position].to_dict(),
+                      '1d':
+                          self._data.loc[self._data.time <= filter_1d, ['figi', position]].groupby(
```
```suggestion
                          self._data.loc[self._data.time >= filter_1d, ['figi', position]].groupby(
```

A typo?

Alexander-Serov replied:
Да, и правда!

sserov replied:
@Alexander-Serov Сломалось и не применилось, кажись

sserov replied:
@Alexander-Serov ?

Alexander-Serov replied:
Should be fixed.

sserov replied:
Fixed in #21

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+        statistics_df.index.name = 'figi'
+        statistics_df['ticker'] = statistics_df.index.map(get_ticker_for_figi)
+        for key, value in statistics.items():
+            if (set(figis) - set(value.keys()) - set(obsolete_tickers.values())):
```
Redundant parentheses? Why does this warning emerge while running? What figis are not present?

sserov replied:
@Alexander-Serov ?

Alexander-Serov replied:
Скобки не нужны, да.
Про фиги вроде уже понятно, нет? Это те, что болше не торгуются.

sserov replied:
Да, окей. Скобки надо убрать. С фигами ясно

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+        msg = [f'\n==\nGood opportunities to buy (criteria: i. 52w change <= '
+               f'{THRESHOLD_MAX_52W_CHG_PERCENT} %, ii. 1w change <= '
+               f'{THRESHOLD_1W_CHG:.2f}):']
+        for figi in statistics_sorted.index:
```
Maybe we can do this without explicit loop by figis, simply by loc for our conditions?

Alexander-Serov replied:
Возможно. Ты опять же можешь предложить.

Alexander-Serov replied:
?

sserov replied:
Думаю, пока забьем, потому что эта статистика --- не основная цель сейчас. Пусть пока останется цикл, если работает. Да и фиджей там пока 47 штук, сильно замедлить не должно

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+            print('Currently no good opportunities to buy :(')
+        # warnings.warn('The output values have not been verified', UserWarning)
+
+    def recommend_other(self, update: bool = True, _print=False, reload=False):
```
Too many code repetition. Needs to be refactored according to TODO

Alexander-Serov replied:
Ок. Все твои контрибьюшнзы будут высоко нами оценены.

Alexander-Serov replied:
?

sserov replied:
Ну тут речь о том, что две функции скопипасчены и просто правила поменяны. Пока забьем по той же причине, что эти рекомендации -- не основная цель

sserov commented on analysis.py:
```diff
@@ -0,0 +1,612 @@
+        THRESHOLD_1D_CHG = -1e-8
+        msg = [f'\n==\nGood opportunities to buy (criteria: i. 52w change <= '
+               f'{THRESHOLD_MAX_52W_CHG_PERCENT} %, ii. previous day change <= '
+               f'{THRESHOLD_1D_CHG:.2f}):']
```
1 day change in the message and in criteria but ```1d_chg_percent``` printed in the second part of the message

Alexander-Serov replied:
И правда. Для недель вроде то же самое.

Alexander-Serov replied:
Исправил.

sserov reviewed (changes requested):
Okay, needs some elaboration and formalization

sserov commented:
@Alexander-Serov

Alexander-Serov committed:
Update README.md

Co-authored-by: Sergey Serov <sserov@users.noreply.github.com>

Alexander-Serov committed:
Update requirements.txt

Co-authored-by: Sergey Serov <sserov@users.noreply.github.com>

Alexander-Serov committed:
Fix nan time bug (#5)

* Never save nan time to the history file.

* Fix logfile name.

* Allow verbose mode for more info.
Fix bug with wrong timestamps in recently downloaded history.

* Change terminal table print precision.

Alexander-Serov commented:
@sserov Кажется, я согласилися на некоторые предложенные изменения, но они почему-то пропали. Если найдешь непримененный, открой снова, пожалуйста.

Alexander-Serov committed:
Apply changes requested in the review. Complete readme and requirements.

Alexander-Serov committed:
Clean up and stanardize.
Update pre-commit config.

Add flake8 checks and standardize.

Add pylint checker.

Alexander-Serov committed:
Add a test for figi history.
Fix dataframe creation from dict, cf https://github.com/Alexander-Serov/tinkoff_analysis/pull/3/files#r444338279

Alexander-Serov committed:
Sm

sserov commented:
Ну вот как раз отвалилось изменение важного косяка с фильтром по 1 дню:

`"1d": self._data.loc[self._data.time <= filter_1d, ["figi", position]]
            .groupby(by="figi")
            .last()[position]
            .to_dict(),`

Alexander-Serov commented:
> Ну вот как раз отвалилось изменение важного косяка с фильтром по 1 дню:
> 
> 
> 
> `"1d": self._data.loc[self._data.time <= filter_1d, ["figi", position]]
> 
>             .groupby(by="figi")
> 
>             .last()[position]
> 
>             .to_dict(),`

Ну что ж, надо будет поправить.

Alexander-Serov commented on requirements.txt:
```diff
@@ -0,0 +1,6 @@
+pandas
```
Меня больше убивает что нельзя этот API который мы используем сюда вставить. Так было бы просто просто устанавливать через пип. ￼

Alexander-Serov replied:
Так тот новый, что ты предложил еще, и можно будет просто через пип устанавливать.
https://github.com/daxartio/tinvest
Создам ишшу #22.

sserov commented on analysis.py:
```diff
@@ -0,0 +1,716 @@
+            merged = old_data.append(new_data, verify_integrity=True, ignore_index=True)
+            # Convert the time column to time because after merge it changes
+            # to object
+            merged.time = pd.to_datetime(self._data.time)
```
A bug! @Alexander-Serov #9 

```suggestion
            merged.time = pd.to_datetime(merged.time)
```

Alexander-Serov replied:
Вот, это мощно, согласен!

sserov reviewed (changes requested):
Added a comment about a bug

Alexander-Serov committed:
Update .gitignore

Co-authored-by: Sergey Serov <sserov@users.noreply.github.com>

Alexander-Serov committed:
Fix >= bug in statistics.

Alexander-Serov committed:
Small changes following review.

sserov commented:
Need to revert 900061a because the code was correct and took the close price of the previous day @Alexander-Serov

Alexander-Serov reviewed (commented):
Fixed following comments.

sserov reviewed (approved):
Everything okay
